mlkem/ntt.c

```diff
@@ -134,18 +197,30 @@ void poly_invntt_tomont(poly *p) { intt_native(p); }
  * Description: Multiplication of polynomials in Zq[X]/(X^2-zeta)
  *              used for multiplication of elements in Rq in NTT domain
  *
+ *              Bounds:
+ *              - a is assumed to be < q in absolute value.
+ *              - Return value < 3/2 q in absolute value
+ *
  * Arguments:   - int16_t r[2]: pointer to the output polynomial
  *              - const int16_t a[2]: pointer to the first factor
  *              - const int16_t b[2]: pointer to the second factor
  *              - int16_t b_cached: Cached precomputation of b[1] * zeta
  **************************************************/
 void basemul_cached(int16_t r[2], const int16_t a[2], const int16_t b[2],
                     int16_t b_cached) {
+  BOUND(a, 2, KYBER_Q, "basemul input bound");
+
   int32_t t0, t1;
+
   t0 = (int32_t)a[1] * b_cached;
   t0 += (int32_t)a[0] * b[0];
   t1 = (int32_t)a[0] * b[1];
   t1 += (int32_t)a[1] * b[0];
+
+  // |ti| < 2 * q * 2^15
```

Comment:
The PQCrystals/Kyber ref/reduce.c code says that the input to montgomery_reduce() must in in
"{-q2^15 .. q2^15 - 1}" but your new code in reduce.c says no constraint on the argument at all.
If the former case is right, then t1 < 2q2^15 doesn't work?

Reply:
But, please feel free to double-check.

Reply:
I'd really like to see a CBMC proof of montgomery_reduce() to prove that it really does work for all int32_t.
I may already have that... let me see...

Reply:
I don't know if CBMC can do that, because it's parametric. I added `test_bounds.py` to brute force this.

Reply:
NOted